frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1495 @@
+			Self::create_snapshot();
+		}
+
+		// for now always start the unsigned phase.
+		<CurrentPhase<T>>::put(Phase::Unsigned((enabled, now)));
+		Self::deposit_event(Event::UnsignedPhaseStarted(Self::round()));
+	}
+
+	/// Creates the snapshot. Writes new data to:
+	///
+	/// 1. [`SnapshotMetadata`]
+	/// 2. [`RoundSnapshot`]
+	/// 3. [`DesiredTargets`]
+	pub fn create_snapshot() {
+		// if any of them don't exist, create all of them. This is a bit conservative.
+		let targets = T::DataProvider::targets();
+		let voters = T::DataProvider::voters();
+		let desired_targets = T::DataProvider::desired_targets();
+
+		<SnapshotMetadata<T>>::put(RoundSnapshotMetadata {
+			voters_len: voters.len() as u32,
+			targets_len: targets.len() as u32,
+		});
+		<DesiredTargets<T>>::put(desired_targets);
+		<Snapshot<T>>::put(RoundSnapshot { voters, targets });
+	}
+
+	/// Checks the feasibility of a solution.
+	///
+	/// This checks the solution for the following:
+	///
+	/// 0. **all** of the used indices must be correct.
+	/// 1. present correct number of winners.
+	/// 2. any assignment is checked to match with [Snapshot::voters].
+	/// 3. for each assignment, the check of `ElectionDataProvider` is also examined.
+	/// 4. the claimed score is valid.
+	fn feasibility_check(
+		solution: RawSolution<CompactOf<T>>,
+		compute: ElectionCompute,
+	) -> Result<ReadySolution<T::AccountId>, FeasibilityError> {
+		let RawSolution { compact, score, round } = solution;
+
+		// first, check round.
+		ensure!(Self::round() == round, FeasibilityError::InvalidRound);
+
+		// winners are not directly encoded in the solution.
+		let winners = compact.unique_targets();
+
+		let desired_targets =
+			Self::desired_targets().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		// NOTE: this is a bit of duplicate, but we keep it around for veracity. The unsigned path
+		// already checked this in `unsigned_per_dispatch_checks`. The signed path *could* check it
+		// upon arrival.
+		ensure!(winners.len() as u32 == desired_targets, FeasibilityError::WrongWinnerCount,);
+
+		// read the entire snapshot.
+		let RoundSnapshot { voters: snapshot_voters, targets: snapshot_targets } =
+			Self::snapshot().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		// ----- Start building. First, we need some closures.
+		let cache = helpers::generate_voter_cache::<T>(&snapshot_voters);
+		let voter_at = helpers::voter_at_fn::<T>(&snapshot_voters);
+		let target_at = helpers::target_at_fn::<T>(&snapshot_targets);
+		let voter_index = helpers::voter_index_fn_usize::<T>(&cache);
+
+		// first, make sure that all the winners are sane.
+		let winners = winners
+			.into_iter()
+			.map(|i| target_at(i).ok_or(FeasibilityError::InvalidWinner))
+			.collect::<Result<Vec<T::AccountId>, FeasibilityError>>()?;
+
+		// Then convert compact -> Assignment. This will fail if any of the indices are gibberish.
+		// that winner indices are already checked.
+		let assignments = compact
```

Comment:
So if we do this first, and then extract the winners from it, it will be a tiny bit faster because: 

Now we query all the winners from `target_at` (which is cheap already) to build all winners. Nonetheless, all the winners are re-queried whilst building assignment. 

Not sure if it is significant enough.

Reply:
having a note just above winners saying it could be potentially optimized is I think a good way to keep that in mind without burden.
I'm also not sure if it is significant.

Reply:
Final remark: 

```
// OPTIMIZATION: we could first build the assignments, and then extract the winners directly
// from that, as that would eliminate a little bit of duplicate work. For now, we keep them
// separate: First extract winners separately from compact, and then assignments. This is
// also better, because we can reject solutions that don't meet `desired_targets` early on.
```